### DISCUSSION TITLE
Arduino OTA + atmega1284p + w5500 lite ethernet

### DISCUSSION THREAD

**Discussion by uzivatel457** [JAndrassy/ArduinoOTA#126] [Q&A]
Hi, I want to upload sketch via ethernet cable.  But I do not understand what to do with this library. And how to upload it on Win10. My theory - need bootloader and main program with reset function. I use atmega1284p and W5500 lite, connected PB4-SS, PB5-MOSI, PB6-MISO, PB7-SCK. Anybody can explain to beginner?

**Comment by JAndrassy** [comment]
read the README.
https://github.com/jandrassy/ArduinoOTA/blob/master/README.md

**Reply by uzivatel457** [reply]
I know about this readme file and i tried to understand many times. but I can not find all my questions... it is not written for beginners step by step.  OTEthernet.ino is main program or bootloader, it looks like different than another bootloader code. 


for example older lib for w5100 and m328  https://sowerbutts.com/optiboot-w5100/ has created hex files. There is main file (bootloader) for compile. And before restart over internet , it set up ethernet boot flag for 10 minutes uploading. And IP config was in EEPROM.

**Comment by JAndrassy** [comment]
mozem slovensky?

**Reply by uzivatel457** [reply]
není problém - jsem CZ :)

**Comment by JAndrassy** [comment]
Aj ked terminologia je anglicka a mozno by to pomohlo aj niekomu inemu. Tak idem anglicky.

The ArduinoOTA libraries (esp8266 and esp32 have their own) are for self update. The Arduino sketch (using this library) writes the received binary to a free space in the flash and then uses some way to copy the binary to the run location of the flash and self reset. So it is not a bootloader. The OTEthernet example shows how to use the library in a sketch.

For classic ATmega only the bootloader can write to flash. So the ArduinoOTA library needs a support in bootloader. The bootloader in MightyCore has this support for ATmega1284p. I guess you use the MightyCore for your ATmega1284p board and you put a bootloader on it using the MightyCore. If yes, then you already may have the right bootloader.

If you tested the Ethernet module and it works and you have the right bootloader, then you can upload the OTEthernet example and wait a little until the Arduino IDE shows the network port in Tools->Port menu (sometimes restarting the IDE helps). After the network port shows, select it and upload the OTEthernet example again. This time it will upload over network.

**Reply by uzivatel457** [reply]
Ok, that's new for me about writing the flash. I have MightyCore - bootloader optiboot_flash_atmega1284p_UART0_115200_16000000L_B0_BIGBOOT.hex and fuses L 0xFF, H 0xD6, E0xFF. So i can upload sketch via arduino IDE - Board MightyCore - 1284, 1284P, programmer stk500 as isp. It is working for usb uploading
But on serial i get 
15:48:51.801 -> Failed to conInitialize Ethernet with DHCP:
15:48:52.702 -> Failed to conInitialize Ethernet with DHCP:
1) it do not get IP address, 2) something restart arduino itself so message on serial line is not complete. In fuses is watchdog disabled
I added   wdt_disable(); in sketch, and still rebooting 

When it get IP address, I will try to look into programming ports.

**Comment by JAndrassy** [comment]
make the Ethernet work first. the WebClient example has enhanced diagnostics

**Comment by uzivatel457** [comment]
I uploaded working project with this ethernet shield. And communication is working. I set up static IP   Ethernet.begin(mac, ip, myDns, gateway, subnet);. Arduino IDE found this device correctly on 192.168.0.195. But during uploading it require password. 
I spotted that  ArduinoOTA.begin(Ethernet.localIP(), "Arduino", "password", InternalStorage); Arduino is with upper case A in Arduino. "password" is not correct
in console i see 
Sketch uses 34946 bytes (26%) of program storage space. Maximum is 130048 bytes.
Global variables use 2414 bytes (14%) of dynamic memory, leaving 13970 bytes for local variables. Maximum is 16384 bytes.
{tools.arduino_ota.cmd} -address 192.168.0.195 -port {upload.network.port} -sketch C:\Users\Pavel\AppData\Local\Temp\arduino_build_452283/IL0211.ino.hex -upload {upload.network.endpoint_upload} -sync {upload.network.endpoint_sync} -reset {upload.network.endpoint_reset} -sync_exp {upload.network.sync_return} 
java.io.IOException: Cannot run program "{tools.arduino_ota.cmd}": CreateProcess error=2, The system cannot find the file specified
so parameters is not written into uploading script

**Comment by JAndrassy** [comment]
did you copy platform.local.txt from extras/avr next to platform.txt in MightyCore?

https://github.com/jandrassy/ArduinoOTA#troubleshooting

**Reply by uzivatel457** [reply]
i just found some time for this project. I copied text from platform.local.txt to platform.txt
when I keep DHCP settings in example code. it do not get IP address (Failed to configure Ethernet using DHCP)
So i set up static IP:
byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xAA, 0xED};
IPAddress ip(192, 168, 0, 85);
IPAddress myDns(192, 168, 0, 254);
IPAddress gateway(192, 168, 0, 254);
IPAddress subnet(255, 255, 255, 0);
EthernetServer server(23);
EthernetClient clients[8];
Ethernet.init(10);  
Ethernet.begin(mac, ip, myDns, gateway, subnet);
> ping to 192.168.0.85 is working, but upload sketch not

processing.app.debug.RunnerException: Unable to connect to 192.168.0.85
	at cc.arduino.packages.uploaders.SSHUploader.uploadUsingPreferences(SSHUploader.java:148)
	at cc.arduino.UploaderUtils.upload(UploaderUtils.java:77)
	at processing.app.SketchController.upload(SketchController.java:732)
	at processing.app.SketchController.exportApplet(SketchController.java:703)
	at processing.app.Editor$UploadHandler.run(Editor.java:2055)
	at java.lang.Thread.run(Thread.java:748)
Unable to connect to 192.168.0.85
